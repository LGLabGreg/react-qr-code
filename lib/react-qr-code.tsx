import React from 'react';

import { ReactQRCodeOptions } from './types';

export const ReactQRCode = (options: ReactQRCodeOptions) => {
  const { data } = options;
  return <>{data}</>;
};
