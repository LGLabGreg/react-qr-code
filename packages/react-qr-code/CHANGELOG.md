# @lglab/react-qr-code

## 1.1.0

### Minor Changes

- 27140cf: Add download api
