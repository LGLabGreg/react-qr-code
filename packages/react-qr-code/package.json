{
  "name": "@lglab/react-qr-code",
  "version": "1.4.2",
  "description": "React library to generate QR codes",
  "author": "LGLab",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/LGLabGreg/react-qr-code.git",
    "directory": "packages/react-qr-code"
  },
  "homepage": "https://reactqrcode.com",
  "keywords": [
    "react",
    "react-component",
    "react-library",
    "qr",
    "code",
    "qrcode",
    "qr-code",
    "qr-code-generator",
    "qr-code-library",
    "qr-code-react",
    "customizable-qr-code",
    "qr-code-with-logo",
    "qr-code-svg",
    "qr-code-image",
    "qr-code-download",
    "typescript",
    "open-source"
  ],
  "type": "module",
  "scripts": {
    "build": "tsc -b && vite build"
  },
  "types": "./dist/index.d.ts",
  "main": "./dist/index.es.js",
  "module": "./dist/index.es.js",
  "exports": {
    ".": {
      "import": "./dist/index.es.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": [
    "dist"
  ],
  "peerDependencies": {
    "react": "^18 || ^19"
  },
  "devDependencies": {
    "@types/node": "^22.13.4",
    "@types/react": "^19.0.8",
    "@types/react-dom": "^19.0.4",
    "@vitejs/plugin-react": "^4.3.4",
    "globals": "^15.15.0",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "typescript": "~5.7.3",
    "typescript-eslint": "^8.24.1",
    "vite": "^6.1.0",
    "vite-plugin-dts": "^4.5.0"
  }
}
