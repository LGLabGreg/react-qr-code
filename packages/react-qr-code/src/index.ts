export type { ReactQRCodeProps } from './types';
export { ReactQRCode } from './react-qr-code';
