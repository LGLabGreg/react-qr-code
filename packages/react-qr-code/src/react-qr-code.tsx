import { useMemo } from 'react';

import { FinderPatternsOuter } from './components/finder-patterns-outer';
import {
	DEFAULT_BGCOLOR,
	DEFAULT_LEVEL,
	DEFAULT_MINVERSION,
	DEFAULT_SIZE,
} from './constants';
import { useQRCode } from './hooks/use-qr-code';
import type { ReactQRCodeProps } from './types/lib';
import { generateDataModulesPath } from './utils/data-modules';

import { FinderPatternsInner } from './components/finder-patterns-inner';
import { excavateModules } from './utils/qr-code';
import {
	sanitizeDataModulesSettings,
	sanitizeFinderPatternInnerSettings,
	sanitizeFinderPatternOuterSettings,
} from './utils/settings';

const ReactQRCode = (props: ReactQRCodeProps) => {
	const {
		ref,
		value,
		size = DEFAULT_SIZE,
		level = DEFAULT_LEVEL,
		bgColor = DEFAULT_BGCOLOR,
		minVersion = DEFAULT_MINVERSION,
		boostLevel,
		marginSize,
		imageSettings,
		svgProps,
	} = props;

	const dataModulesSettings = useMemo(
		() => sanitizeDataModulesSettings(props.dataModulesSettings),
		[props.dataModulesSettings],
	);

	const finderPatternOuterSettings = useMemo(
		() => sanitizeFinderPatternOuterSettings(props.finderPatternOuterSettings),
		[props.finderPatternOuterSettings],
	);

	const finderPatternInnerSettings = useMemo(
		() => sanitizeFinderPatternInnerSettings(props.finderPatternInnerSettings),
		[props.finderPatternInnerSettings],
	);

	const { margin, cells, numCells, calculatedImageSettings } = useQRCode({
		value,
		level,
		minVersion,
		boostLevel,
		marginSize,
		imageSettings,
		size,
	});

	let modules = cells;
	let image = null;
	if (imageSettings != null && calculatedImageSettings != null) {
		if (calculatedImageSettings.excavation != null) {
			modules = excavateModules(cells, calculatedImageSettings.excavation);
		}

		image = (
			<image
				href={imageSettings.src}
				height={calculatedImageSettings.h}
				width={calculatedImageSettings.w}
				x={calculatedImageSettings.x + margin}
				y={calculatedImageSettings.y + margin}
				preserveAspectRatio="none"
				opacity={calculatedImageSettings.opacity}
				// Note: specified here always, but undefined will result in no attribute.
				crossOrigin={calculatedImageSettings.crossOrigin}
			/>
		);
	}

	const dataModulesPath = generateDataModulesPath({
		modules,
		margin,
		settings: dataModulesSettings,
	});

	return (
		<svg
			height={size}
			width={size}
			viewBox={`0 0 ${numCells} ${numCells}`}
			ref={ref}
			role="img"
			aria-label={svgProps?.['aria-label'] || 'QR Code'}
			{...svgProps}
		>
			<path fill={bgColor} d={`M0,0 h${numCells}v${numCells}H0z`} />
			<FinderPatternsOuter
				modules={modules}
				margin={margin}
				settings={finderPatternOuterSettings}
			/>
			<FinderPatternsInner
				modules={modules}
				margin={margin}
				settings={finderPatternInnerSettings}
			/>
			<path
				fill={dataModulesSettings.color}
				d={dataModulesPath}
				shapeRendering="crispEdges"
			/>
			{image}
		</svg>
	);
};

ReactQRCode.displayName = 'ReactQRCode';

export { ReactQRCode };
