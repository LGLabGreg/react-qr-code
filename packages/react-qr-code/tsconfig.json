{
  "compilerOptions": {
    "target": "ES2020",
    "module": "ESNext",
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "declaration": true,
    "declarationMap": true,
    "sourceMap": true,
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "react-jsx",
    "noEmit": true,
    "outDir": "./dist",
    "types": ["@testing-library/jest-dom"]
  },
  "include": ["src/**/*"],
  "exclude": ["dist", "node_modules"]
}
