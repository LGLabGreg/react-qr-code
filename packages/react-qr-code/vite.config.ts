import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'

const __dirname = dirname(fileURLToPath(import.meta.url))

// https://vite.dev/config/
export default defineConfig({
  plugins: [
    dts({
      rollupTypes: true,
    }),
  ],
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'ReactQRCode',
      fileName: (format) => `index.${format}.js`,
      formats: ['es'],
    },
    rollupOptions: {
      external: ['react', 'react-dom', 'react/jsx-runtime'],
      output: {
        globals: {
          react: 'React',
          'react-dom': 'ReactDOM',
        },
      },
    },
  },
})
