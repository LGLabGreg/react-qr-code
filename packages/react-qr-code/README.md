![React QR Code Header](https://github.com/LGLabGreg/react-qr-code/raw/main/media/repo-header.png)

# @lglab/react-qr-code

A highly customizable and lightweight QR code generator for React applications.

ğŸš€ **Live Demo:** [reactqrcode.com/demo](https://reactqrcode.com/demo)  
ğŸ“– **Documentation:** [reactqrcode.com](https://reactqrcode.com/)

## Quick Features

- **Highly Customizable** - Style the finder patterns, modules, and colors exactly how you want.
- **Performance Optimized** - Generates QR codes efficiently without sacrificing quality.
- **SVG-Based Rendering** - Crisp and scalable output for web and print
- **Developer-Friendly** - Built with TypeScript, easy to use, and well-documented.

## Install

```
pnpm add @lglab/react-qr-code
```

or

```
npm i @lglab/react-qr-code
```

or

```
yarn add @lglab/react-qr-code
```

or

```
bun add @lglab/react-qr-code
```

## Quick Start

```
import { ReactQRCode } from '@lglab/react-qr-code'

const Page = () => {
  return (
    <ReactQRCode value='https://reactqrcode.com' />
  );
}
```
