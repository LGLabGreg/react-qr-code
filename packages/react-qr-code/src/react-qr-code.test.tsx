import { render } from '@testing-library/react'
import { expect, test } from 'vitest'

import { ReactQRCode } from './react-qr-code'

test('renders QR code component', () => {
  const result = render(<ReactQRCode value='test' />)
  expect(result.container).toBeInTheDocument()
  expect(result).toMatchSnapshot()
})
