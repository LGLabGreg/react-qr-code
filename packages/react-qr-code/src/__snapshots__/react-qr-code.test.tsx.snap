// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders QR code component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <svg
        aria-label="QR Code"
        height="128"
        role="img"
        viewBox="0 0 29 29"
        width="128"
      >
        <path
          d="M0,0 h29v29H0z"
          fill="#FFFFFF"
        />
        <path
          d="M 4 4v 7h 7v -7zM 5 5h 5v 5h -5zM 18 4v 7h 7v -7zM 19 5h 5v 5h -5zM 4 18v 7h 7v -7zM 5 19h 5v 5h -5z"
          fill="#000000"
        />
        <rect
          fill="#000000"
          height="3"
          rx="0"
          width="3"
          x="6"
          y="6"
        />
        <rect
          fill="#000000"
          height="3"
          rx="0"
          width="3"
          x="20"
          y="6"
        />
        <rect
          fill="#000000"
          height="3"
          rx="0"
          width="3"
          x="6"
          y="20"
        />
        <path
          d="M12,4h1v1h-1ZM13,4h1v1h-1ZM14,4h1v1h-1ZM16,4h1v1h-1ZM13,5h1v1h-1ZM14,5h1v1h-1ZM14,6h1v1h-1ZM12,7h1v1h-1ZM15,7h1v1h-1ZM16,7h1v1h-1ZM14,8h1v1h-1ZM15,8h1v1h-1ZM13,9h1v1h-1ZM15,9h1v1h-1ZM12,10h1v1h-1ZM14,10h1v1h-1ZM16,10h1v1h-1ZM13,11h1v1h-1ZM16,11h1v1h-1ZM6,12h1v1h-1ZM8,12h1v1h-1ZM9,12h1v1h-1ZM10,12h1v1h-1ZM12,12h1v1h-1ZM13,12h1v1h-1ZM17,12h1v1h-1ZM21,12h1v1h-1ZM24,12h1v1h-1ZM7,13h1v1h-1ZM8,13h1v1h-1ZM11,13h1v1h-1ZM12,13h1v1h-1ZM16,13h1v1h-1ZM18,13h1v1h-1ZM19,13h1v1h-1ZM23,13h1v1h-1ZM24,13h1v1h-1ZM4,14h1v1h-1ZM5,14h1v1h-1ZM6,14h1v1h-1ZM7,14h1v1h-1ZM8,14h1v1h-1ZM10,14h1v1h-1ZM12,14h1v1h-1ZM16,14h1v1h-1ZM17,14h1v1h-1ZM18,14h1v1h-1ZM19,14h1v1h-1ZM21,14h1v1h-1ZM22,14h1v1h-1ZM23,14h1v1h-1ZM24,14h1v1h-1ZM4,15h1v1h-1ZM5,15h1v1h-1ZM6,15h1v1h-1ZM7,15h1v1h-1ZM9,15h1v1h-1ZM12,15h1v1h-1ZM13,15h1v1h-1ZM15,15h1v1h-1ZM16,15h1v1h-1ZM19,15h1v1h-1ZM20,15h1v1h-1ZM23,15h1v1h-1ZM4,16h1v1h-1ZM5,16h1v1h-1ZM9,16h1v1h-1ZM10,16h1v1h-1ZM12,16h1v1h-1ZM13,16h1v1h-1ZM14,16h1v1h-1ZM19,16h1v1h-1ZM21,16h1v1h-1ZM23,16h1v1h-1ZM24,16h1v1h-1ZM12,17h1v1h-1ZM14,17h1v1h-1ZM15,17h1v1h-1ZM16,17h1v1h-1ZM21,17h1v1h-1ZM24,17h1v1h-1ZM13,18h1v1h-1ZM14,18h1v1h-1ZM20,18h1v1h-1ZM21,18h1v1h-1ZM23,18h1v1h-1ZM24,18h1v1h-1ZM12,19h1v1h-1ZM13,19h1v1h-1ZM14,19h1v1h-1ZM16,19h1v1h-1ZM18,19h1v1h-1ZM23,19h1v1h-1ZM12,20h1v1h-1ZM14,20h1v1h-1ZM15,20h1v1h-1ZM18,20h1v1h-1ZM20,20h1v1h-1ZM21,20h1v1h-1ZM23,20h1v1h-1ZM24,20h1v1h-1ZM13,21h1v1h-1ZM18,21h1v1h-1ZM19,21h1v1h-1ZM23,21h1v1h-1ZM12,22h1v1h-1ZM16,22h1v1h-1ZM17,22h1v1h-1ZM19,22h1v1h-1ZM22,22h1v1h-1ZM24,22h1v1h-1ZM13,23h1v1h-1ZM14,23h1v1h-1ZM15,23h1v1h-1ZM17,23h1v1h-1ZM18,23h1v1h-1ZM19,23h1v1h-1ZM20,23h1v1h-1ZM21,23h1v1h-1ZM23,23h1v1h-1ZM14,24h1v1h-1ZM17,24h1v1h-1ZM18,24h1v1h-1ZM19,24h1v1h-1ZM22,24h1v1h-1ZM23,24h1v1h-1ZM24,24h1v1h-1Z"
          fill="#000000"
          shape-rendering="crispEdges"
        />
      </svg>
    </div>
  </body>,
  "container": <div>
    <svg
      aria-label="QR Code"
      height="128"
      role="img"
      viewBox="0 0 29 29"
      width="128"
    >
      <path
        d="M0,0 h29v29H0z"
        fill="#FFFFFF"
      />
      <path
        d="M 4 4v 7h 7v -7zM 5 5h 5v 5h -5zM 18 4v 7h 7v -7zM 19 5h 5v 5h -5zM 4 18v 7h 7v -7zM 5 19h 5v 5h -5z"
        fill="#000000"
      />
      <rect
        fill="#000000"
        height="3"
        rx="0"
        width="3"
        x="6"
        y="6"
      />
      <rect
        fill="#000000"
        height="3"
        rx="0"
        width="3"
        x="20"
        y="6"
      />
      <rect
        fill="#000000"
        height="3"
        rx="0"
        width="3"
        x="6"
        y="20"
      />
      <path
        d="M12,4h1v1h-1ZM13,4h1v1h-1ZM14,4h1v1h-1ZM16,4h1v1h-1ZM13,5h1v1h-1ZM14,5h1v1h-1ZM14,6h1v1h-1ZM12,7h1v1h-1ZM15,7h1v1h-1ZM16,7h1v1h-1ZM14,8h1v1h-1ZM15,8h1v1h-1ZM13,9h1v1h-1ZM15,9h1v1h-1ZM12,10h1v1h-1ZM14,10h1v1h-1ZM16,10h1v1h-1ZM13,11h1v1h-1ZM16,11h1v1h-1ZM6,12h1v1h-1ZM8,12h1v1h-1ZM9,12h1v1h-1ZM10,12h1v1h-1ZM12,12h1v1h-1ZM13,12h1v1h-1ZM17,12h1v1h-1ZM21,12h1v1h-1ZM24,12h1v1h-1ZM7,13h1v1h-1ZM8,13h1v1h-1ZM11,13h1v1h-1ZM12,13h1v1h-1ZM16,13h1v1h-1ZM18,13h1v1h-1ZM19,13h1v1h-1ZM23,13h1v1h-1ZM24,13h1v1h-1ZM4,14h1v1h-1ZM5,14h1v1h-1ZM6,14h1v1h-1ZM7,14h1v1h-1ZM8,14h1v1h-1ZM10,14h1v1h-1ZM12,14h1v1h-1ZM16,14h1v1h-1ZM17,14h1v1h-1ZM18,14h1v1h-1ZM19,14h1v1h-1ZM21,14h1v1h-1ZM22,14h1v1h-1ZM23,14h1v1h-1ZM24,14h1v1h-1ZM4,15h1v1h-1ZM5,15h1v1h-1ZM6,15h1v1h-1ZM7,15h1v1h-1ZM9,15h1v1h-1ZM12,15h1v1h-1ZM13,15h1v1h-1ZM15,15h1v1h-1ZM16,15h1v1h-1ZM19,15h1v1h-1ZM20,15h1v1h-1ZM23,15h1v1h-1ZM4,16h1v1h-1ZM5,16h1v1h-1ZM9,16h1v1h-1ZM10,16h1v1h-1ZM12,16h1v1h-1ZM13,16h1v1h-1ZM14,16h1v1h-1ZM19,16h1v1h-1ZM21,16h1v1h-1ZM23,16h1v1h-1ZM24,16h1v1h-1ZM12,17h1v1h-1ZM14,17h1v1h-1ZM15,17h1v1h-1ZM16,17h1v1h-1ZM21,17h1v1h-1ZM24,17h1v1h-1ZM13,18h1v1h-1ZM14,18h1v1h-1ZM20,18h1v1h-1ZM21,18h1v1h-1ZM23,18h1v1h-1ZM24,18h1v1h-1ZM12,19h1v1h-1ZM13,19h1v1h-1ZM14,19h1v1h-1ZM16,19h1v1h-1ZM18,19h1v1h-1ZM23,19h1v1h-1ZM12,20h1v1h-1ZM14,20h1v1h-1ZM15,20h1v1h-1ZM18,20h1v1h-1ZM20,20h1v1h-1ZM21,20h1v1h-1ZM23,20h1v1h-1ZM24,20h1v1h-1ZM13,21h1v1h-1ZM18,21h1v1h-1ZM19,21h1v1h-1ZM23,21h1v1h-1ZM12,22h1v1h-1ZM16,22h1v1h-1ZM17,22h1v1h-1ZM19,22h1v1h-1ZM22,22h1v1h-1ZM24,22h1v1h-1ZM13,23h1v1h-1ZM14,23h1v1h-1ZM15,23h1v1h-1ZM17,23h1v1h-1ZM18,23h1v1h-1ZM19,23h1v1h-1ZM20,23h1v1h-1ZM21,23h1v1h-1ZM23,23h1v1h-1ZM14,24h1v1h-1ZM17,24h1v1h-1ZM18,24h1v1h-1ZM19,24h1v1h-1ZM22,24h1v1h-1ZM23,24h1v1h-1ZM24,24h1v1h-1Z"
        fill="#000000"
        shape-rendering="crispEdges"
      />
    </svg>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
